unison-cli/src/Unison/Codebase/Editor/HandleInput/Update.hs

```diff
@@ -189,12 +201,324 @@ handleUpdate input optionalPatch requestedNames = do
+        --   #pingpong.pong => ("fresh2", <"fresh1" + 2>)
+        --   #wham          => ("fresh3", <"fresh2" + 3>)
+        let refToGeneratedNameAndTerm :: Map TermReferenceId (Symbol, Term Symbol Ann)
+            refToGeneratedNameAndTerm =
+              -- Running example:
+              --
+              --   #pingpong.pong => (<#pingpong.ping + 2>, { "pong" })
+              --   #wham          => (<#pingpong.pong + 3>, { "wham" })
+              implicitTerms
+                -- Running example:
+                --
+                --   #pingpong.pong => <#newping + 2>
+                --   #wham          => <#pingpong.pong + 3>
+                & Map.map (\(term, _names) -> rewrite term)
+                -- Running example:
+                --
+                --   #newping       => <#wham + 4>
+                --   #pingpong.pong => <#newping + 2>
+                --   #wham          => <#pingpong.pong + 3>
+                & Map.union interimRefToTerm
+                & Term.unhashComponent
+              where
+                -- Running example:
+                --
+                --   #newping => <#wham + 4>
+                interimRefToTerm :: Map TermReferenceId (Term Symbol Ann)
+                interimRefToTerm =
+                  Map.remap (\(_var, (ref, _wk, term, _typ)) -> (ref, term)) nameToInterimInfo
+                -- Running example: apply the following reference mapping everwhere in a term:
+                --
+                --   #pingpong.ping -> #newping
+                --   ref            -> ref
+                rewrite :: Term Symbol Ann -> Term Symbol Ann
+                rewrite =
+                  updatedNameToOldRefs
+                    & Map.toList
+                    & foldMap
+                      ( \(name, oldRefs) ->
+                          oldRefs
+                            & Set.toList
+                            & map
+                              ( \oldRef ->
+                                  case bijLookupL name nameToInterimRef of
+                                    Nothing -> undefined
+                                    Just interimRef -> (oldRef, interimRef)
+                              )
+                            & Map.fromList
+                      )
+                    & rewriteTermReferences
+
+        let unisonFile :: UnisonFile Symbol Ann
+            unisonFile =
+              UnisonFileId
+                { dataDeclarationsId = UF.dataDeclarationsId' (Slurp.originalFile slurp0),
+                  effectDeclarationsId = UF.effectDeclarationsId' (Slurp.originalFile slurp0),
+                  -- Running example:
+                  --
+                  --   fresh1 = fresh3 + 4
+                  --   fresh2 = fresh1 + 2
+                  --   fresh3 = fresh2 + 3
+                  terms = Map.elems refToGeneratedNameAndTerm,
+                  -- In the context of this update, whatever watches were in the latest typechecked Unison file are
```

Comment:
What happens to updates of test watches which are part of a component? E.g. If ping & pong were tests and the edit to `ping` happened inside of a test watch?

Reply:
> If ping & pong were tests and the edit to ping happened inside of a test watch?

You can't perform an `update` on a watched term